Objects/stringlib/fastsearch.h

```diff
@@ -324,222 +359,402 @@ typedef struct STRINGLIB(_pre) {
+    const STRINGLIB_CHAR *p = pw->p;
+    Py_ssize_t m = pw->m;
+    Py_ssize_t stt = dir == 1 ? 0 : m - 1;
+    Py_ssize_t end = dir == 1 ? m - 1 : 0;
+    Py_ssize_t not_found_shift = Py_MIN(m, MAX_SHIFT);
+    for (Py_ssize_t i = 0; i < (Py_ssize_t)TABLE_SIZE; i++) {
+        pw->table[i] = Py_SAFE_DOWNCAST(not_found_shift,
+                                        Py_ssize_t, SHIFT_TYPE);
     }
-    else {
-        // A lower bound on the period
-        p->period = Py_MAX(p->cut, len_needle - p->cut) + 1;
+    for (Py_ssize_t i = m - not_found_shift; i < m; i++) {
+        SHIFT_TYPE shift = Py_SAFE_DOWNCAST(m - 1 - i,
+                                            Py_ssize_t, SHIFT_TYPE);
+        pw->table[p[stt + dir*i] & TABLE_MASK] = shift;
     }
-    // The gap between the last character and the previous
-    // occurrence of an equivalent character (modulo TABLE_SIZE)
-    p->gap = len_needle;
-    STRINGLIB_CHAR last = needle[len_needle - 1] & TABLE_MASK;
-    for (Py_ssize_t i = len_needle - 2; i >= 0; i--) {
-        STRINGLIB_CHAR x = needle[i] & TABLE_MASK;
+    // 2. Initialize "Good Suffix" Last Character Gap
+    // Note: gap("___aa") = 1
+    pw->gap = m;
+    STRINGLIB_CHAR last = p[end] & TABLE_MASK;
+    for (Py_ssize_t i = 1; i < m; i++) {
+        STRINGLIB_CHAR x = p[end - dir*i] & TABLE_MASK;
         if (x == last) {
-            p->gap = len_needle - 1 - i;
+            pw->gap = i;
             break;
         }
     }
-    // Fill up a compressed Boyer-Moore "Bad Character" table
-    Py_ssize_t not_found_shift = Py_MIN(len_needle, MAX_SHIFT);
-    for (Py_ssize_t i = 0; i < (Py_ssize_t)TABLE_SIZE; i++) {
-        p->table[i] = Py_SAFE_DOWNCAST(not_found_shift,
-                                       Py_ssize_t, SHIFT_TYPE);
+    LOG("Good Suffix Gap: %ld\n", pw->gap);
+}
+
+
+static void
+STRINGLIB(_init_critical_fac)(STRINGLIB(prework) *pw, int dir)
+{
+    // Calculate Critical Factorization
+    const STRINGLIB_CHAR *p = pw->p;
+    Py_ssize_t m = pw->m;
+    Py_ssize_t cut, period;
+    int is_periodic;
+    cut = STRINGLIB(_factorize)(p, m, &period, dir);
+    assert(cut + period <= m);
+    if (dir == 1) {
+        is_periodic = 0 == memcmp(p, p + period, cut * STRINGLIB_SIZEOF_CHAR);
     }
-    for (Py_ssize_t i = len_needle - not_found_shift; i < len_needle; i++) {
-        SHIFT_TYPE shift = Py_SAFE_DOWNCAST(len_needle - 1 - i,
-                                            Py_ssize_t, SHIFT_TYPE);
-        p->table[needle[i] & TABLE_MASK] = shift;
+    else {
+        Py_ssize_t cut_idx = m - cut;
+        is_periodic = (0 == memcmp(p + cut_idx, p + cut_idx - period,
+                                   cut * STRINGLIB_SIZEOF_CHAR));
+    }
+    if (is_periodic) {
+        assert(cut <= m/2);
+        assert(cut < period);
+        LOG("Needle is periodic.\n");
     }
+    else {
+        // A lower bound on the period
+        // CLARIFY> An upper bound?
```

Comment:
A comment is a "lower bound". Shouldn't it be upper bound? Just double checking if I am not missing some POV.